rpc/client/httpclient.go

```diff
@@ -48,15 +49,53 @@ func NewHTTP(remote, wsEndpoint string) *HTTP {
 	return &HTTP{
 		rpc:      rc,
 		remote:   remote,
+		batching: false,
 		WSEvents: newWSEvents(cdc, remote, wsEndpoint),
 	}
 }
 
 var _ Client = (*HTTP)(nil)
 
+// StartBatch will cause the HTTP client to go into batching mode, which will
+// enqueue subsequent requests (`c.Status()`, `c.ABCIQuery`,
+// `c.BroadcastTxSync`, etc.) instead of sending them immediately. These
+// requests will be enqueued until such time that they are either sent by
+// calling `c.SendBatch()`, or cleared by calling `c.CancelBatch()`.
+func (c *HTTP) StartBatch() {
```

Comment:
Maybe add a comment that StartBatch isn't thread-safe/not meant to be used from go-rountines?

Reply:
Agree. Was our client thread-safe before? Can't find any documentation. If yes, then we should try to preserve this property.

In any case, we should document whenever it's thread-safe to use or not (for both HTTP and Local clients).

Reply:
I've refactored the batching functionality to be, IMHO, a little more user-friendly when it comes to concurrent usage. This would allow for multiple batches to be created across different goroutines, all executed by the same client without interfering with each other.

The new usage would look something like:

```go
c := client.NewHTTP("127.0.0.1:26657", "/websocket")

// Generate two random key/value pairs and their equivalent transactions
tx1 := makeTxKV()
tx2 := makeTxKV()

// Create a new batch
batch := c.NewBatch()

// Queue up the first transaction
batch.BroadcastTxCommit(tx1)
// Queue up the second transaction
batch.BroadcastTxCommit(tx2)

// Send the whole batch
batch.Send()
```